### DISCUSSION TITLE
[Jetbrains plugin] Feedback on essential features missing to make Copilot more useful for my team

### DISCUSSION THREAD

**Discussion by cmp-nct** [community/community#36052] [Copilot Conversations]
I have multiple requests/reports in here, it would be appreciated if a Plugin Dev can take a look.  
I tested on stable and nightly

**1) Clear and quick signal if Copilot is searching, if not when pressing the right keys for it.**
One of the main issues I have is that Copilot does not signal in any way if it is working, waiting, idle.
I keep pressing the key for "next suggestion" or "give suggestion" and nothing happens.
Sometimes I get a warning "no completions available" and when pressing 2 more times one comes up.
So I spend a significant amount of time trying to make it "react", that really does not work well yet.
**When requesting a completion copilot needs to be more intuitive and visualize it's status properly.**

**2) Control over how much I want to complete when working in existing code !**
In many cases I want to rework a snippet or a single line, I also want to get a completion for ONE line or for a whole snippet.
In the case of "snippets" I often get one line after another and in case of one-line I often get a whole snippet afterward.
**We need shortcuts to request: "one line" "multi-line" "full functionality". Especially "one line" is important**
Imagine this  simplified case:
```
if ($x == 'a' || $x == 'b' || or $x == 'c') 
{
   // lots of existing code
   // ...
```
I want copilot to change this into in_array(), the best way to do this currently is to move comment the if(), move to the next line and start.
`if (in_array($x, `
This works well except for the problem that copilot now wants to add not only the correct new if() case but it also wants to write the entire content inside the {} which already exists and works.
Especially given that the code already exists this ruins the code, you need to be very careful to remove excessive lines after 'Tab'

_A quick solution is to get a "one line completion" shortcut. 
Longterm I would like to see a "replace selected code" or "look at selected code and selected comment".
So we can tell Copilot to look at one snippet with increased 'weights' and then do something with it, optimize it, rewrite it._

3) Copilot should show a confidence score with results, so we know how well it understood it.
When requesting completions Copilot is not capable of (not enough code to learn from) then you usually get very strange results.
I tested requesting machine learning models in PHP, there are only 2 big projects out and those are not that frequently used yet.
Copilot brought up quite interesting snippets, all of them wrong but a starting point.
They also included written communication from some discussion forum, bug reports about the code, questions "help me please" and answers from moderators closing those. 
It's not a must-have but it would help to see "copilot confidence score: very low" next to a snippet. So you have to be more careful.

4.1) When the plugin detects that a SQL query is needed, add the structure of the database as context, if possible give it a CREATE TABLE syntax as well. Of course the whole DB is often not enough, though the used tables in the code before and after give an indication, the comments and names give an indication and it would be possible to ask copilot a double question:
a) What table should I use for (snippet of sql-indicative code to complete/work on) + meta information (all table names in the current database(es)
b) copilot answers internally, likely very quick: "table X, Y"
c) now ask for the actual completion and include TABLE CREATE syntax
4.2) Do the same trick with project class structures, for large projects: get the most used classes and methods into context, for small projects provide a prototype class with all the public variables. Provide that alongside the actual requests so Copilot can give answers depending on available code.

5) Create a complimentary "project" and "file" based meta document we can access, read and even write.
This should be passed alongside the usual code to copilot (of course we are limited in lines/words)
So copilot could be fed permanently with a context.
I have projects with tens of thousands of lines per file, so copilot has no chance to work with it.
If I had access to a "meta document" per project I would pass 5-10 lines of explanations there.
On a file based meta document I would keep that up-to-date with the most relevant stuff I am working on right now.
I'd add function usage examples in there, SQL create syntax, maybe 2-3 lines in a comment explaining what I am working on (right now).

Suggestions 4) and 5) would make copilot a lot smarter in how it responds, it would give more power to the developers who want to inform their copilot about the route.

**Comment by iamcjwilliams** [comment]
@comp-nect Thank you for your feedback, our copilot dev team has been notified!

**Comment by ossanna16** [comment]
Thank you for the great feedback @cmp-nct! We appreciate it 🚀

**Reply by Huntercry** [reply]
Ntp, ya dejé de confiar
